### PR TITLE
Added support for parsing certificates with RSAES-OAEP public keys

### PR DESCRIPTION
**Pull request by DenisKarch**
This will currently return an *rsa.PublicKey throwing away the hashing
and masking algorithms.
We are working with go upstream to find a more permanent solution.

### CODE REVIEW

File: x509/x509_test.go [hunk 1]
```diff
@@ -1119,6 +1119,35 @@ func TestRSAPSSSelfSigned(t *testing.T) {
...
...065006d0065006e00743081a10603551d2304819930819680148ffd47880e239a3a3a20de13edf101e882a9d21da17ba4793077310b3009060355040613024445310f300d060355040813065361786f6e793121301f060355040a1318496e66696e656f6e20546563686e6f6c6f67696573204147310c300a060355040b130341494d312630240603550403131d4946582054504d20454b20496e7465726d6564696174652043412032308201053081930603551d0904818b308188303a06035504343133300b300906052b0e03021a05003024302206092a864886f70d0101073015a213301106092a864886f70d010109040454435041301606056781050210310d300b0c03312e32020102020103303206056781050212312930270101ffa0030a0101a1030a0100a2030a0100a310300e1603332e310a01040a01010101ff0101ff300d06092a864886f70d010105050003820101003b25d3958cf08b2c32cefb40570f845e25fa98fd38a47be8c16e4ef663d8f057824ff550052b4a338e353b08bfe70364cda07961d3f4fb5cbf548497389a5ab59c3469d14b6f5bfdc27db3dd0cbd21ac818a64a14032cd433a2bffb939b5f54555c2b6892b5f6479e3c38bd4b30283d2e8ee57ecad779dae90bd9d6052ad6bf3efbb30f639e03b0a239f539a476b129540bc9806e2dc7011d265dca8a95301aaba872b4e176bdb670e65b750b0afdc2ca97b3f0bef55862ae135bd86b1d3a28d4eedfe5ee445a3f1d65bbad0adae49999e613eabb133166461cf47146de078a7d7f550940a1a475ecc834ee784a6a6e42fa4ad7869ef8d2b2251c830efb38496`
```
Comment (ericchiang) [inline x509/x509_test.go]: nit: consider PEM encoding this so it's not just a super long string. https://golang.org/pkg/encoding/pem/

Also, explain where this cert comes from, and why we care about parsing it.

File: x509/x509_test.go [hunk 2]
```diff
@@ -1119,6 +1119,35 @@ func TestRSAPSSSelfSigned(t *testing.T) {
...
+
+func TestParseCertificateWithRSAESOAEPPublicKey(t *testing.T) {
+	expectedKey := &rsa.PublicKey{
```
Comment (daviddrysdale) [inline x509/x509_test.go]: nit: s/expected/want/

File: x509/x509.go [hunk 3]
```diff
@@ -238,6 +244,37 @@ const (
 	SHA512WithRSAPSS
 )
 
+// RFC 4055
+// Various identifiers
```
Comment (daviddrysdale) [inline x509/x509.go]: Do these constants need to be exported?

File: x509/x509.go [hunk 4]
```diff
@@ -1272,6 +1314,41 @@ func parsePublicKey(algo PublicKeyAlgorithm, keyData *publicKeyInfo, nfe *NonFat
...
+		}
+		if len(rest) != 0 {
+			return nil, errors.New("x509: trailing data after RSA public key")
+		}
+		if p.N.Sign() <= 0 {
+			nfe.AddError(errors.New("x509: RSA modulus is not a positive number"))
+		}
+		if p.E <= 0 {
+			return nil, errors.New("x509: RSA public exponent is not a positive number")
+		}
```
Comment (daviddrysdale) [inline x509/x509.go]: Could do with a comment here pointing out that we're checking the `Parameters` format but otherwise ignoring/dropping it.

Comment (daviddrysdale) [reply x509/x509.go]: By the way, is the public key usable without the parameter values?  (I don't know anything about RSAES-OAEP)  Should there be a `TODO(someone): export parameter values` or a `TODO(someone): replace this functionality with whatever upstream `crypto/x509` ends up supporting` ?

Comment (DenisKarch) [reply x509/x509.go]: Technically no it shouldn't be. You need the parameters for the RSAES-OAEP encryption scheme. The key itself is still just rsa and can be used as such but it really shouldn't be (in my opinion it would be best to modify either the PublicKeyAlgorithm in the Certificate struct to hold the parameters relevant to the scheme, but that might be considered an aggressive change). I'm still not sure what upstream will do so I've left off on making any such changes. I've added a TODO.

File: x509/x509_test.go (hunk 1 above)
Comment (DenisKarch) [reply x509/x509_test.go]: I couldn't find any easy way to convert the cert from DER to PEM (probably because of the RSAES-OAEP public key).

Comment (ericchiang) [reply x509/x509_test.go]: PEM is just base64'ing the DER data. Here's an example https://play.golang.org/p/-ADspaRm91-

Let me know if you have any other questions :)

File: x509/x509_test.go [hunk 5]
```diff
@@ -1119,6 +1119,39 @@ func TestRSAPSSSelfSigned(t *testing.T) {
 	}
 }
 
+// Valid EKCert (from a TPM)  with RSAES-OAEP Public Key.
```
Comment (ericchiang) [inline x509/x509_test.go]: Which hardware vendor is this? Can you include that information in your comment?

File: x509/x509.go [hunk 6]
```diff
@@ -1264,14 +1329,14 @@ func parsePublicKey(algo PublicKeyAlgorithm, keyData *publicKeyInfo, nfe *NonFat
 		if len(rest) != 0 {
 			return nil, errors.New("x509: trailing data after RSA public key")
 		}
-
```
Comment (daviddrysdale) [inline x509/x509.go]: nit: you've removed a blank line here (sorry to be pedantic, but any unnecessary changes make upstream merges harder)

File: x509/x509.go [hunk 7]
```diff
@@ -1244,12 +1294,27 @@ type distributionPointName struct {
...
-	case RSA:
+	case RSA, RSAESOAEP:
 		// RSA public keys must have a NULL in the parameters.
 		// See RFC 3279, Section 2.3.1.
-		if !bytes.Equal(keyData.Algorithm.Parameters.FullBytes, asn1.NullBytes) {
+		if algo == RSA && !bytes.Equal(keyData.Algorithm.Parameters.FullBytes, asn1.NullBytes) {
 			nfe.AddError(errors.New("x509: RSA key missing NULL parameters"))
 		}
+		if algo == RSAESOAEP {
+			// We only parse the parameters to ensure it is a valid encoding, we throw out the actually values
```
Comment (daviddrysdale) [inline x509/x509.go]: nit: add a spec ref here, s/actually/actual/

File: x509/x509.go (hunk 4 above)
Comment (daviddrysdale) [reply x509/x509.go]: Is there an upstream issue we can track (and mention here)?

Comment (DenisKarch) [reply x509/x509.go]: I've mostly been talking with upstream internally, there is an open issue and I've linked it.
